hoomd/mpcd/ReverseNonequilibriumShearFlow.cc

```diff
@@ -0,0 +1,276 @@
+    {
+    m_p_target = p_target;
+    }
+
+/*!
+ * \param slab_width Slab thickness
+ */
+void mpcd::ReverseNonequilibriumShearFlow::setSlabWidth(Scalar slab_width)
+    {
+    // check that delta is bigger than zero
+    if (slab_width <= Scalar(0))
+        {
+        throw std::runtime_error("invalid slab thickness in ReverseNonequilibriumShearFlow");
+        }
+
+    m_slab_width = slab_width;
+    requestUpdateSlabs();
+    }
+
+/*!
+ * \param slab_separation Slab separation
+ */
+void mpcd::ReverseNonequilibriumShearFlow::setSlabSeparation(Scalar slab_separation)
+    {
+    // check that delta is bigger than zero
+    if (slab_separation <= Scalar(0))
+        {
+        throw std::runtime_error("invalid slab separation in ReverseNonequilibriumShearFlow");
+        }
+
+    m_slab_separation = slab_separation;
+    requestUpdateSlabs();
+    }
+
+void mpcd::ReverseNonequilibriumShearFlow::setSlabs()
+    {
+    const BoxDim& global_box = m_pdata->getGlobalBox();
+    const Scalar3 global_lo = global_box.getLo();
+    const Scalar3 global_hi = global_box.getHi();
+
+    const Scalar lo_pos = -m_slab_separation * Scalar(0.5);
+    const Scalar hi_pos = m_slab_separation * Scalar(0.5);
+    const Scalar delta_half = Scalar(0.5) * m_slab_width;
+    m_hi_pos = make_scalar2(hi_pos - delta_half, hi_pos + delta_half);
+    m_lo_pos = make_scalar2(lo_pos - delta_half, lo_pos + delta_half);
+
+    // explicit check that the two regions for velocity swapping are inside
+    // of the box entirely and don't overlap
+    if (m_hi_pos.y >= global_hi.y || m_lo_pos.x < global_lo.y || m_lo_pos.y > m_hi_pos.x)
+        {
+        throw std::runtime_error("Invalid slab configuration in ReverseNonequilibriumShearFlow");
+        }
+    }
+
+/*!
+ * \param timestep Current time step of the simulation
+ */
+void mpcd::ReverseNonequilibriumShearFlow::update(uint64_t timestep)
+    {
+    // if slabs have changed, update them
+    if (m_update_slabs)
+        {
+        setSlabs();
+        m_update_slabs = false;
+        }
+
+    findSwapParticles();
+    swapPairMomentum();
+    }
+
+/*!
+ * Finds all particles in the "min" and "max" slab in y direction and
+ * puts them into two GPUArrays, sorted by their momentum closest to p_target in x direction.
+ */
+void mpcd::ReverseNonequilibriumShearFlow::findSwapParticles()
+    {
+    // get needed array handles. All in read mode, no particle is changed in this search function
+    ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(),
+                               access_location::host,
+                               access_mode::read);
+    ArrayHandle<Scalar4> h_pos(m_mpcd_pdata->getPositions(),
+                               access_location::host,
+                               access_mode::read);
+
+    // sort pairs according to their sign and momentum
+    // temporary pairs to hold particles in slab
+    // first index is momentum in x dir - second index is particle tag
```

Comment:
Try to keep comments up to date.
```suggestion
    // sort pairs according to their momentum
```
Noting that I would like to come back and replace this by a `GPUArray` and `std::partial_sort`.